### PR TITLE
fix promise bugs on corner cases (first one)

### PR DESCRIPTION
**Pull request by stibear**
Ref: #339

### CODE REVIEW

File: contrib/20.r7rs/scheme/lazy.scm [hunk 1]
```diff
@@ -6,32 +6,51 @@
...
+	      (promise-value promise)
+	      (begin
+		(set-promise-done!  promise (promise-done? new-promise))
+		(set-promise-value! promise (promise-value new-promise))
+		(set-box! new-promise (unbox promise))
+		(force promise))))))
 
   (define (make-promise obj)
-    (if (promise? obj)
+    (if (and (pair? obj) (promise? obj))
```
Comment (KeenS) [inline contrib/20.r7rs/scheme/lazy.scm]: [nits] please define `box?` and use it for consistency

**Comment by KeenS** [comment]
It seems that all the call of `make-promise%` followed by `box` so why don't  you define the combination as an procedure?

**Comment by nyuichi** [comment]
@stibear 

Can you split the patch into two things, one for the first bug and the other for the second one?
I can merge the first one immediately, while the second needs more discussion, I think.

File: contrib/20.r7rs/scheme/lazy.scm (hunk 1 above)
Comment (stibear) [inline contrib/20.r7rs/scheme/lazy.scm]: corrected & recommitted

**Comment by nyuichi** [comment]
Can we implement delay-force just like the following instead of using boxes?:

```
  (define-syntax delay-force
    (syntax-rules ()
      ((_ expr)
       (let ((p (make-promise% #f #f)))
         (set-promise-value! p (lambda () (let ((v (begin expr))) (set-promise-done! p #t) v)))))))
```

**Comment by stibear** [comment]
FAIL: (force np)
   expected ok but got ng

**Comment by nyuichi** [comment]
@stibear 

I made an alternative implementation using tri-state rather than boxes: https://github.com/picrin-scheme/picrin/tree/stibear-lazy

**Comment by stibear** [comment]
@nyuichi It looks no problem.

**Pull request merged**